scipy/optimize/_trustregion_constr/minimize_trustregion_constr.py

```diff
@@ -0,0 +1,485 @@
+                if self.first_iteration:
+                    if np.linalg.norm(delta_x) != 0:
+                        hess.instanciate_matrix(delta_x, delta_grad)
+                        hess.scale_matrix(delta_x, delta_grad)
+                        hess.update(delta_x, delta_grad)
+                        self.first_iteration = False
+                    else:
+                        hess.instanciate_matrix(delta_x, delta_grad)
+                else:
+                    hess.update(delta_x, delta_grad)
+                return hess
+        else:
+            hess_wrapped = None  
+        self.hess = hess_wrapped
+
+
+class sqp_printer:
+    @staticmethod
+    def print_header():
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^10}|{4:^10}|{5:^10}|{6:^10}|"
+              .format("niter", "f evals", "CG iter", "tr radius", "penalty",
+                      "opt", "c viol"))
+        s = "-"*6 + ":"
+        s2 = ":" + "-"*8 + ":"
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^10}|{4:^10}|{5:^10}|{6:^10}|"
+              .format(s, s, s, s2, s2, s2, s2))
+
+    @staticmethod
+    def print_problem_iter(niter, nfev, cg_niter, tr_radius,
+                           penalty, opt, c_viol):
+        print("|{0:>7}|{1:>7}|{2:>7}| {3:^1.2e} | {4:^1.2e} |"
+              " {5:^1.2e} | {6:^1.2e} |"
+              .format(niter, nfev, cg_niter, tr_radius, penalty,
+                      opt, c_viol))
+
+    @staticmethod
+    def print_footer():
+        print("")
+        print((7*3 + 10*4 + 8)*"-")
+        print("")
+
+
+class ip_printer:
+    @staticmethod
+    def print_header():
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^13}|{4:^10}|{5:^10}|{6:^10}|{7:^10}|"
+              .format("niter", "f evals", "CG iter", "barrier param",
+                      "tr radius", "penalty", "opt", "c viol"))
+        s = "-"*6 + ":"
+        s2 = ":" + "-"*11 + ":"
+        s3 = ":" + "-"*8 + ":"
+        print("|{0:^7}|{1:^7}|{2:^7}|{3:^13}|{4:^10}|{5:^10}|{6:^10}|{7:^10}|"
+              .format(s, s, s, s2, s3, s3, s3, s3))
+
+    @staticmethod
+    def print_problem_iter(niter, nfev, cg_niter, barrier_parameter, tr_radius,
+                           penalty, opt, c_viol):
+        print("|{0:>7}|{1:>7}|{2:>7}|   {3:^1.2e}  | "
+              "{4:^1.2e} | {5:^1.2e} | {6:^1.2e} | {7:^1.2e} |"
+              .format(niter, nfev, cg_niter, barrier_parameter,
+                      tr_radius, penalty, opt, c_viol))
+
+    @staticmethod
+    def print_footer():
+        print("")
+        print((7*3 + 13 + 10*4 + 9)*"-")
+        print("")
+
+
+def _minimize_trustregion_constr(fun, x0, args, grad,
+                                 hess, hessp, constraints,
+                                 xtol=1e-8, gtol=1e-8,
+                                 barrier_tol=1e-8,
+                                 sparse_jacobian=None,
+                                 callback=None, maxiter=1000,
+                                 verbose=0, finite_diff_options=None,
```

Comment:
I'm confused with `finite_diff_options`. It looks unnecessary here and unclear how to use it.

Reply:
The idea is that whenever you set ``grad = '2-point'`` or ``hess = '2-point'``, you make use of finite difference routine you have implemented: ``approx_derivative``.

This is a dictionary containing options for this function. Possible options would include:
``rel_step``; ``sparsity``, ...

So it is not unnecessary. But I agree with you that it is not very well documented, I will try to improve the description of this parameter a little bit.

Reply:
I'm confused, I thought you defined finite difference options for different objects separately, like:
```
jac : {callable, ‘2-point’, ‘3-point’, ‘cs’, bool}, optional
```

Reply:
Ok, it is indeed very confusing: I pass the options ‘2-point’, ‘3-point’, ‘cs’ as you said, but the remaining options I am currently reading from the dictionary ``finite_diff_options``.

I think the best options is to include all finite difference arguments (e.g. ``rel_step``; ``sparsity``, ...) as options to the solver and remove this dictionary ``finite_diff_options``.